Websocket failure on future3

Hey Just fount out there's a websocket failure after installing future 3. Any Idea of what could be causing this?

This is the error found on the dev console on Chrome:

"webSocketEndpoint.ts:56 WebSocket connection to 'ws://raspberrypi.local:5557/' failed: "

Hardware Raspberry Pi4 8GB

I just tried it @DivineDominion on a fresh 32bit bookworm os lite (`6.1.0-rpi4-rpi-v7l #1 SMP Raspbian 1:6.1.54-1+rpt2 (2023-10-05) armv7l`) and RPI4 and the problem still persists. The only solution was to run 

```bash
ZMQ_DRAFT_API=1 ZMQPREFIX="/usr/local" pip3 install --no-binary=pyzmq --break-system-packages pyzmq
```

i.e. adding the `--break-system-packages` option to the `pip3 install` command as mentioned in https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2050#issuecomment-1774101077 above by @DivineDominion  and @s-martin (I'd prefer setting this option instead of modifying the `pip.conf` file as in https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2050#issuecomment-1774081800 by @DivineDominion). The same option had to be added in `_jukebox_core_install_python_requirements`, see https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/0660586f7ec96aea75006cc8df8b86e1d3d11542/installation/routines/setup_jukebox_core.sh#L111

I also added a fix for the installation of npm as the current installation method has been deprecated (and does not install npm anymore) and there has been an issue with the installation of `pyalsaaudio` during the installation of the `requirements.txt` which is why I had to install `libasound2-dev` via `apt-get`

<details>

Here is the traceback during the invocation of `pip3 install ... -r .../requirements.txt`

```
Installing collected packages: snowballstemmer, pyalsaaudio, pulsectl, evdev, tornado, sphinxcontrib-jsmath, ruamel.yaml.clib, python_mpd2, Pygments, pyflakes, pycodestyle, pluggy, packaging, mock, mccabe, MarkupSafe, iniconfig, imagesize, docutils, babel, alabaster, ruamel.yaml, pytest, Jinja2, flake8, sphinxcontrib-serializinghtml, sphinxcontrib-qthelp, sphinxcontrib-htmlhelp, sphinxcontrib-devhelp, sphinxcontrib-applehelp, sphinx, sphinxcontrib-jquery, sphinx_rtd_theme
  Running setup.py install for pyalsaaudio ... error
  error: subprocess-exited-with-error
  
  × Running setup.py install for pyalsaaudio did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      running install
      /usr/lib/python3/dist-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_ext
      building 'alsaaudio' extension
      creating build
      creating build/temp.linux-armv7l-cpython-311
      arm-linux-gnueabihf-gcc -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/usr/include/python3.11 -c alsaaudio.c -o build/temp.linux-armv7l-cpython-311/alsaaudio.o
      alsaaudio.c:28:10: fatal error: alsa/asoundlib.h: No such file or directory
         28 | #include <alsa/asoundlib.h>
            |          ^~~~~~~~~~~~~~~~~~
      compilation terminated.
      error: command '/usr/bin/arm-linux-gnueabihf-gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> pyalsaaudio

```
</details>

If you like, I can open a PR with these changes, see https://github.com/MiczFlor/RPi-Jukebox-RFID/compare/future3/develop...Chilipp:RPi-Jukebox-RFID:fix-python-install?expand=1 

the installation (after implement the fix for the arm64 OS in https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2041) was successful via

```bash
cd; GIT_USER='Chilipp' GIT_BRANCH='fix-python-install' bash <(wget -qO- https://raw.githubusercontent.com/MiczFlor/RPi-Jukebox-RFID/future3/develop/installation/install-jukebox.sh)
```

and the `jukebox-daemon.service` was running successfully and the webapp has been built

--- a/installation/routines/setup_jukebox_core.sh
+++ b/installation/routines/setup_jukebox_core.sh
@@ -27,6 +27,7 @@ _jukebox_core_install_os_dependencies() {
     python3-rpi.gpio python3-gpiozero \
     espeak ffmpeg mpg123 \
     pulseaudio pulseaudio-module-bluetooth pulseaudio-utils caps \
+    libasound2-dev \
     --no-install-recommends \
     --allow-downgrades \
     --allow-remove-essential \

--- a/installation/routines/setup_jukebox_core.sh
+++ b/installation/routines/setup_jukebox_core.sh
@@ -99,7 +99,7 @@ _jukebox_core_build_and_install_pyzmq() {
     fi
 
     sudo ZMQ_PREFIX="${ZMQ_PREFIX}" ZMQ_DRAFT_API=1 \
-         pip3 install --no-cache-dir --no-binary "pyzmq" --pre pyzmq
+         pip3 install --no-cache-dir --no-binary "pyzmq" --pre --break-system-packages pyzmq
   else
     echo "    Skipping. pyzmq already installed"
   fi
@@ -108,7 +108,7 @@ _jukebox_core_build_and_install_pyzmq() {
 _jukebox_core_install_python_requirements() {
   echo "  Install requirements"
   cd "${INSTALLATION_PATH}"  || exit_on_error
-  sudo pip3 install --no-cache-dir -r "${INSTALLATION_PATH}/requirements.txt"
+  sudo pip3 install --no-cache-dir --break-system-packages -r "${INSTALLATION_PATH}/requirements.txt"
 }
 
 _jukebox_core_install_settings() {

--- a/installation/routines/setup_jukebox_core.sh
+++ b/installation/routines/setup_jukebox_core.sh
@@ -33,6 +33,13 @@ _jukebox_core_install_os_dependencies() {
     --allow-remove-essential \
     --allow-change-held-packages
 
+  # add configuration that we break the global python system packages
+  # (required for bookworm, see 
+  # https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2050). This
+  # should be removed once the jukebox has been isolated to a dedicated
+  # venv
+  sudo python3 -m pip config set global.break-system-packages true
+
   sudo pip3 install --upgrade pip
 }
 
@@ -100,7 +107,7 @@ _jukebox_core_build_and_install_pyzmq() {
     fi
 
     sudo ZMQ_PREFIX="${ZMQ_PREFIX}" ZMQ_DRAFT_API=1 \
-         pip3 install --no-cache-dir --no-binary "pyzmq" --pre --break-system-packages pyzmq
+         pip3 install --no-cache-dir --no-binary "pyzmq" --pre pyzmq
   else
     echo "    Skipping. pyzmq already installed"
   fi
@@ -109,7 +116,7 @@ _jukebox_core_build_and_install_pyzmq() {
 _jukebox_core_install_python_requirements() {
   echo "  Install requirements"
   cd "${INSTALLATION_PATH}"  || exit_on_error
-  sudo pip3 install --no-cache-dir --break-system-packages -r "${INSTALLATION_PATH}/requirements.txt"
+  sudo pip3 install --no-cache-dir -r "${INSTALLATION_PATH}/requirements.txt"
 }
 
 _jukebox_core_install_settings() {

--- a/installation/routines/setup_jukebox_webapp.sh
+++ b/installation/routines/setup_jukebox_webapp.sh
@@ -73,6 +73,9 @@ _jukebox_webapp_register_as_system_service_with_nginx() {
   sudo mv -f /etc/nginx/sites-available/default /etc/nginx/sites-available/default.orig
   sudo cp -f "${INSTALLATION_PATH}/resources/default-settings/nginx.default" /etc/nginx/sites-available/default
 
+  # make sure nginx can access the home directory of the user
+  sudo chmod o+x /home/pi
+
   sudo service nginx restart
 }
 

--- a/installation/routines/setup_jukebox_webapp.sh
+++ b/installation/routines/setup_jukebox_webapp.sh
@@ -4,7 +4,7 @@
 GD_ID_COMPILED_WEBAPP="1EE_1MdneGtKL5V7GyYZC0nb6ODQWTsPb" # https://drive.google.com/file/d/1EE_1MdneGtKL5V7GyYZC0nb6ODQWTsPb/view?usp=sharing
 
 # For ARMv7+
-NODE_SOURCE="https://deb.nodesource.com/setup_16.x"
+NODE_MAJOR=16
 # For ARMv6
 # To update version, follow these links
 # https://github.com/sdesalas/node-pi-zero
@@ -26,12 +26,18 @@ _jukebox_webapp_install_node() {
     # Zero and older versions of Pi with ARMv6 only
     # support experimental NodeJS
     if [[ $(uname -m) == "armv6l" ]]; then
-      NODE_SOURCE=${NODE_SOURCE_EXPERIMENTAL}
+      wget -O - ${NODE_SOURCE_EXPERIMENTAL} | sudo bash
+      sudo apt-get -qq -y install nodejs
+      sudo npm install --silent -g npm
+    else
+      # install NodeJS and npm as recommended in
+      # https://github.com/nodesource/distributions
+      curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | sudo gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
+      echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR.x nodistro main" | sudo tee /etc/apt/sources.list.d/nodesource.list
+      sudo apt-get update
+      sudo apt-get install nodejs npm -y
     fi
 
-    wget -O - ${NODE_SOURCE} | sudo bash
-    sudo apt-get -qq -y install nodejs
-    sudo npm install --silent -g npm
   fi
 }
 

